src/Swift.Runtime/src/Swift/Runtime/InteropServices/SwiftMarshal.cs

```diff
@@ -15,22 +18,31 @@ public static class SwiftMarshal
     /// </summary>
     /// <typeparam name="T">The type of the value being marshaled</typeparam>
     /// <param name="value">The value to marshal</param>
-    /// <param name="swiftDest">the destination for marshaling</param>
-    /// <returns>A pointer to memory to pass in to Swift for marshaling. Note: this value may be different from the value passed in.</returns>
+    /// <param name="swiftDestSpan">the destination for marshaling</param>
     /// <exception cref="NotSupportedException"></exception>
-    public static IntPtr MarshalToSwift<T>(T value, IntPtr swiftDest)
+    public static void MarshalToSwift<T>(T value, Span<byte> swiftDestSpan)
     {
         if (value is ISwiftObject swiftValue)
         {
-            return swiftValue.MarshalToSwift(new IntPtr(swiftDest));
+            swiftValue.MarshalToSwift(swiftDestSpan);
+            return;
         }
 
         var type = typeof(T);
-        if (type.IsPrimitive || typeof(nint).IsAssignableFrom(type) || typeof(nuint).IsAssignableFrom(type))
+        if ((type.IsPrimitive || typeof(nint).IsAssignableFrom(type) || typeof(nuint).IsAssignableFrom(type)) && !typeof(char).IsAssignableFrom(type))
         {
             unsafe
             {
-                return new IntPtr(MarshalPrimitiveToSwift(value, (void*)swiftDest));
+                int size = Unsafe.SizeOf<T>();
+                if (size != swiftDestSpan.Length)
```

Comment:
This is overly aggressive. It is entirely reasonable for a default amount of memory to be used during marshalling. Instead of a strict size requirement I would just ensure it is sufficient.

`if (size > swiftDestSpan.Length)`

This might also be an opportunity to return return the size in bytes of the marshalling operation.

Reply:
Thanks, updated. `SwiftMarshal` now returns size in bytes of the marshalling operation.